### PR TITLE
[Runtime] NFC: Quiet global ctor warnings

### PR DESCRIPTION
**Pull request by davezarzycki**
Review request: please note the added `FIXME`.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

File: stdlib/public/runtime/Metadata.cpp [hunk 1]
```diff
@@ -3288,8 +3292,11 @@ static GenericWitnessTableCache &getCache(GenericWitnessTable *gen) {
                 sizeof(GenericWitnessTable::PrivateDataType),
                 "metadata cache is larger than the allowed space");
 
+  SWIFT_ALLOWED_RUNTIME_GLOBAL_CTOR_BEGIN
+  // FIXME -- Why is this Lazy<T> but not static?
```
Comment (gottesmm) [inline stdlib/public/runtime/Metadata.cpp]: Can you file an SR for this as well?

**Review by gottesmm — COMMENTED** [review]
I am not the best person to review this, but a quick drive by.

File: stdlib/public/runtime/Metadata.cpp (hunk 1 above)
Comment (davezarzycki) [reply stdlib/public/runtime/Metadata.cpp]: SR-7619

**Comment by davezarzycki** [comment]
For reference, `Lazy<T>` generates global-ctor warnings because of how `std::once_flag` is implemented in libcxx.

**Comment by rjmccall** [comment]
> For reference, `Lazy<T>` generates global-ctor warnings because of how `std::once_flag` is implemented in libcxx.

Really?  `std::once_flag`'s constructor should be `constexpr`, which has been a requirement since its introduction in C++11.  Is there something weird in your configuration?

File: stdlib/public/runtime/Metadata.cpp (hunk 1 above)
Comment (rjmccall) [reply stdlib/public/runtime/Metadata.cpp]: The generic witness table's private data object is emitted by the compiler.

What warnings are you suppressing here?  Nothing about this code generates a global constructor, or for that matter a global object.

Comment (rjmccall) [reply stdlib/public/runtime/Metadata.cpp]: Note that this code is basically assuming that Lazy<T> constant-initializes to a zero bit-pattern.

**Comment by davezarzycki** [comment]
I don't think that anything is noteworthy about my setup except that I regularly build swift with top-of-tree clang/llvm/lld and I use unified builds for building Swift. Otherwise, this is a standard Red Hat Fedora 28 install with the OS provided libcxx from LLVM (version 6.0).

I just forced `CMAKE_CXX_FLAGS` to have `-std=c++14` set, but no change in warning output. If I remember correctly when I looked into this, I think the "is trivially constructible" type trait fails for `std::once_flag`.

**Comment by rjmccall** [comment]
I would probably start with preprocessed source and try to reduce it.  Just sanity-check that `constexpr` is actually in all the places it should be.

**Comment by davezarzycki** [comment]
I figured it out. This probably is a clang bug. Do you agree?

If one deletes `T t` or `Once once` below, then no warning is produced. If both exist, then `declaration requires a global constructor [-Wglobal-constructors]` erroneously fires.
```
struct Once {
    unsigned value;
    constexpr Once() : value(0) {}
};

template<typename T>
class Lazy {
    T t;
    Once once;
public:
    constexpr Lazy() = default;
};
Lazy<unsigned> lazy;
```

**Comment by davezarzycki** [comment]
I found a one line workaround.

@swift-ci please smoke test

**Comment by rjmccall** [comment]
That's really weird; could you file a PR at bugs.llvm.org?  I wonder if there's some nonsense thing about exceptions at play here.

Your workaround looks good.

**Review by rjmccall — APPROVED** [review]

**Comment by davezarzycki** [comment]
https://bugs.llvm.org/show_bug.cgi?id=37362

**Pull request merged**